domains/client/domain-operator/src/domain_bundle_producer.rs

```diff
@@ -211,6 +237,23 @@ where
                 .propose_bundle_at(proof_of_election, self.parent_chain.clone(), tx_range)
                 .await?;
 
+            let receipt_has_no_extrinsics = bundle_header
+                .receipt
```

Comment:
Instead of just one ER, I think we need to check all domain blocks that are still in the challenge period, namely the domain block whose `extrinsics_root` is still reachable through `ExecutionInbox`.

It is more convenient to check `extrinsics_root` through the runtime state `ExecutionInbox` than checking the local ER's `extrinsics_root` because we don't need to handle a potential consensus fork in runtime. In the runtime, it is also important to check `ExecutionInbox` instead of ER in the block tree because ER may not be up to date, especially after processing fraud proof which may revert some ER.

Reply:
> Instead of just one ER, I think we need to check all domain blocks that are still in the challenge period, namely the domain block whose extrinsics_root is still reachable through ExecutionInbox.

Why? We are only concerned about whther to submit a bundle that is part of the bundle header. I'm not I sure I see any reason to check all the bundles that are in the challenge phase. Please explain in detail why we need to do that ?

Reply:
The only reason we need empty bundle is to accelerate the confirmation of non-empty domain blocks in challenge phase. That being said if there is any non-empty domain block in challenge phase we need to produce bundle to move the challenge period forward.

Reply:
> The only reason we need empty bundle is to accelerate the confirmation of non-empty domain blocks in challenge phase. That being said if there is any non-empty domain block in challenge phase we need to produce bundle to move the challenge period forward.

I know that. But why do we need to check all the bundles that are in Challenge period when we produce a bundle that contains the ER of the parent. As long as the bundle has extrinsics or ER, parent block contain extrinsics, we submit the bundle. I'm not able to imagine the case you are suggesting. Can you elaborate pls ?

Reply:
Sure, assuming the block tree contains domain blocks from `N - BlockTreePruningDepth` to `N`, all the blocks in `N - BlockTreePruningDepth`..`N - 1` is non-empty and block `#N` is empty.

Now the operator tries to produce a bundle that contains the ER of block `#N` but because the tx pool is empty and the block `#N` is empty, it decides to skip producing this bundle, and as long as the tx pool is empty it will perform the same check and keep skipping bundle. As a result, all the non-empty blocks in `N - BlockTreePruningDepth`..`N - 1` are left to non-confirm.

Reply:
Okay what I had in mind was that operators will not confirm ERs with new domain blocks so that they go out of challenge period. 
I think you are suggesting that operators should produce a bundle until all the ERs in the challenge period contains only empty extrinsincs.

My main concern here is, a single transaction will produce `BlockTreePruningDepth` number of domain blocks.

I dont mind changing this but I think we should check with Chen and other. Can you please update the spec and notify pls ?

Reply:
> Okay what I had in mind was that operators will not confirm ERs with new domain blocks so that they go out of challenge period.

Emmm, the only way to derive ER go out of challenge period is to produce `BlockTreePruningDepth` number of domain blocks, that is our security assumption.

> I think you are suggesting that operators should produce a bundle until all the ERs in the challenge period contains only empty extrinsincs.
>
> My main concern here is, a single transaction will produce BlockTreePruningDepth number of domain blocks.

Yes, but the consequence if we don't do this is a transaction can left to non-confirm potentially forever.

> I dont mind changing this but I think we should check with Chen and other. Can you please update the spec and notify pls ?

Sure, will do!